### ISSUE TITLE
ScopedDiagnosticHandler supports derivation but lacks a virtual destructor

### ISSUE THREAD

**Issue by stellaraccident** [llvm/llvm-project#59212]
There are situations where we parameterize the precise selection of diagnostic handler in tools, and the lack of a virtual destructor means that doing something like the following can leak memory:

```
std::unique_ptr<ScopedDiagnosticHandler> diagnosticHandler;
if (doSomething) {
  diagnosticHandler = std::make_unique<SourceMgrDiagnosticHandler>();
}

// Do stuff under the diagnostic handler.
```

While perhaps a bit of an abuse of the API, it is perfectly legal and we should add an inline virtual destructor.

**Comment by joker-eph** [comment]
There is no virtual table associated with `ScopedDiagnosticHandler` right now, it is fairly common that we don't support what you're doing here to avoid adding virtual table unnecessarily (and avoid spurious indirection in destruction, which in turns prevents inlining, etc).

That said this class is rarely enough used and not on any critical path that it is probably fine. Just worth keeping in mind that the "legality" of what you're doing is IMO really a choice we have and not automatically derived from "it supports derivation".

**Comment by stellaraccident** [comment]
Yeah, I had a similar thought process. It is rare/not hot so should be fine.

However, I feel like this class hierarchy mixes some concerns and probably should have been done differently: I would have expected that the RAII "Scoped" support would be a template that was parameterized over a concrete diagnostic handler (vs mixing the scoping and the specific functionality via subclassing).

**Comment by stellaraccident** [comment]
(also thanks for having a look: I was in the middle of something and was just noting the issue so as not to lose it vs expecting you to pick it up)

**Comment by lattner** [comment]
Agreed, I don't see any harm to having a virtual destructor here.  If there is utility in it, we should just add it!

**Comment by stellaraccident** [comment]
https://reviews.llvm.org/D138764
